// appium-android-driver/build/test/functional/helpers.js
'use strict';

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumAdb = require('appium-adb');

var _appiumAdb2 = _interopRequireDefault(_appiumAdb);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _appiumSupport = require('appium-support');

var MOCHA_TIMEOUT = process.env.MOCHA_TIMEOUT ? process.env.MOCHA_TIMEOUT : process.env.TRAVIS ? 120000 : 15000;

var CHROMEDRIVER_2_20_ASSET_MAP = {
  windows: ['windows', 'chromedriver.exe'],
  mac: ['mac', 'chromedriver'],
  linux32: ['linux-32', 'chromedriver'],
  linux64: ['linux-64', 'chromedriver']
};

function getChromedriver220Asset() {
  var basePath, dir, cmd, _CHROMEDRIVER_2_20_ASSET_MAP$windows, _CHROMEDRIVER_2_20_ASSET_MAP$mac, _CHROMEDRIVER_2_20_ASSET_MAP;

  return _regeneratorRuntime.async(function getChromedriver220Asset$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        basePath = _path2['default'].resolve(__dirname, '..', '..', '..', 'test', 'assets', 'chromedriver-2.20');
        dir = undefined;
        cmd = undefined;

        if (!_appiumSupport.system.isWindows()) {
          context$1$0.next = 9;
          break;
        }

        _CHROMEDRIVER_2_20_ASSET_MAP$windows = _slicedToArray(CHROMEDRIVER_2_20_ASSET_MAP.windows, 2);
        dir = _CHROMEDRIVER_2_20_ASSET_MAP$windows[0];
        cmd = _CHROMEDRIVER_2_20_ASSET_MAP$windows[1];
        context$1$0.next = 23;
        break;

      case 9:
        if (!_appiumSupport.system.isMac()) {
          context$1$0.next = 15;
          break;
        }

        _CHROMEDRIVER_2_20_ASSET_MAP$mac = _slicedToArray(CHROMEDRIVER_2_20_ASSET_MAP.mac, 2);
        dir = _CHROMEDRIVER_2_20_ASSET_MAP$mac[0];
        cmd = _CHROMEDRIVER_2_20_ASSET_MAP$mac[1];
        context$1$0.next = 23;
        break;

      case 15:
        context$1$0.next = 17;
        return _regeneratorRuntime.awrap(_appiumSupport.system.arch());

      case 17:
        context$1$0.t0 = context$1$0.sent;
        context$1$0.t1 = 'linux' + context$1$0.t0;
        context$1$0.t2 = CHROMEDRIVER_2_20_ASSET_MAP[context$1$0.t1];
        _CHROMEDRIVER_2_20_ASSET_MAP = _slicedToArray(context$1$0.t2, 2);
        dir = _CHROMEDRIVER_2_20_ASSET_MAP[0];
        cmd = _CHROMEDRIVER_2_20_ASSET_MAP[1];

      case 23:
        return context$1$0.abrupt('return', _path2['default'].resolve(basePath, dir, cmd));

      case 24:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function ensureAVDExists(mochaContext, avdName) {
  var adb;
  return _regeneratorRuntime.async(function ensureAVDExists$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(_appiumAdb2['default'].createADB());

      case 2:
        adb = context$1$0.sent;
        context$1$0.prev = 3;
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(adb.checkAvdExist(avdName));

      case 6:
        context$1$0.next = 12;
        break;

      case 8:
        context$1$0.prev = 8;
        context$1$0.t0 = context$1$0['catch'](3);

        mochaContext.skip();
        return context$1$0.abrupt('return', false);

      case 12:
        return context$1$0.abrupt('return', true);

      case 13:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[3, 8]]);
}

exports.MOCHA_TIMEOUT = MOCHA_TIMEOUT;
exports.ensureAVDExists = ensureAVDExists;
exports.getChromedriver220Asset = getChromedriver220Asset;
