// appium-youiengine-driver/build/lib/commands/find.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumBaseDriver = require('appium-base-driver');

var _utils = require('../utils');

var commands = {};

commands.findElOrEls = function callee$0$0(strategy, selector, mult, context) {
  var createGetElementCommand, result, doFind;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    var _this = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        createGetElementCommand = function createGetElementCommand(strategy, selector, mult, context) {

          if (typeof context === "undefined" || !context) {
            context = '';
          }

          var ext = mult ? 's' : '';
          var commandObject = {
            args: ['' + selector, '' + context]
          };

          switch (strategy) {
            case "name":
              commandObject.name = 'getElement' + ext + 'ByName';
              break;
            case "accessibility id":
              commandObject.name = 'getElement' + ext + 'ByName';
              break;
            case "id":
              commandObject.name = 'getElement' + ext + 'ById';
              break;
            case "class name":
              commandObject.name = 'getElement' + ext + 'ByType';
              break;
            default:
              throw new Error("Unsupported strategy for findElOrEls: " + strategy);
          }

          return JSON.stringify(commandObject);
        };

        result = undefined;

        doFind = function doFind() {
          var findByAxIdCmd, res;
          return _regeneratorRuntime.async(function doFind$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                findByAxIdCmd = createGetElementCommand(strategy, selector, mult, context);
                context$2$0.next = 3;
                return _regeneratorRuntime.awrap(this.executeSocketCommand(findByAxIdCmd));

              case 3:
                res = context$2$0.sent;
                context$2$0.prev = 4;

                result = JSON.parse(res);
                context$2$0.next = 11;
                break;

              case 8:
                context$2$0.prev = 8;
                context$2$0.t0 = context$2$0['catch'](4);
                throw new Error("Bad response from findElOrEls");

              case 11:
                if (!(result.status === _utils.youiEngineDriverReturnValues.WEBDRIVER_INVALID_SELECTOR)) {
                  context$2$0.next = 13;
                  break;
                }

                throw new _appiumBaseDriver.errors.InvalidSelectorError(result.value);

              case 13:
                if (!(result.value === "")) {
                  context$2$0.next = 15;
                  break;
                }

                return context$2$0.abrupt('return', false);

              case 15:
                return context$2$0.abrupt('return', true);

              case 16:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this, [[4, 8]]);
        };

        context$1$0.prev = 3;
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(this.implicitWaitForCondition(doFind));

      case 6:
        context$1$0.next = 15;
        break;

      case 8:
        context$1$0.prev = 8;
        context$1$0.t0 = context$1$0['catch'](3);

        if (!(context$1$0.t0.message && context$1$0.t0.message.match(/Condition unmet/))) {
          context$1$0.next = 14;
          break;
        }

        throw new _appiumBaseDriver.errors.NoSuchElementError(result.value);

      case 14:
        throw context$1$0.t0;

      case 15:
        return context$1$0.abrupt('return', result.value);

      case 16:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[3, 8]]);
};

exports['default'] = commands;
module.exports = exports['default'];

// parse error

// looks like we have to check the status or resulting value to see that it's not empty.

// condition was not met, throw NoSuchElementError

// some other issue occurred, report it
