// appium-base-driver/build/lib/basedriver/commands/timeout.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var _asyncbox = require('asyncbox');

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumSupport = require('appium-support');

var _protocol = require('../../protocol');

var _driver = require("../driver");

var _driver2 = _interopRequireDefault(_driver);

var commands = {},
    helpers = {},
    extensions = {};

var MIN_TIMEOUT = 0;

// If we define `commands.timeouts` instead of `commands.timeoutsW3C`, the command `timeouts` will be called
// from other dirver's timeouts. See https://github.com/appium/appium-base-driver/pull/164
// Arguments will be: [{"protocol":"W3C","implicit":30000}, "1dcfe021-8fc8-49bd-8dac-e986d3091b97", ...]
// eslint-disable-next-line no-unused-vars
commands.timeouts = function callee$0$0(timeoutsObj) {
  var script, pageLoad, implicit, type, ms;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(timeoutsObj.protocol === _driver2['default'].DRIVER_PROTOCOL.W3C)) {
          context$1$0.next = 16;
          break;
        }

        script = timeoutsObj.script;
        pageLoad = timeoutsObj.pageLoad;
        implicit = timeoutsObj.implicit;

        _logger2['default'].debug('script: ' + script + ', pageLoad: ' + pageLoad + ', implicit: ' + implicit);

        if (!_appiumSupport.util.hasValue(script)) {
          context$1$0.next = 8;
          break;
        }

        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(this.scriptTimeoutW3C(script));

      case 8:
        if (!_appiumSupport.util.hasValue(pageLoad)) {
          context$1$0.next = 11;
          break;
        }

        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(this.pageLoadTimeoutW3C(pageLoad));

      case 11:
        if (!_appiumSupport.util.hasValue(implicit)) {
          context$1$0.next = 14;
          break;
        }

        context$1$0.next = 14;
        return _regeneratorRuntime.awrap(this.implicitWaitW3C(implicit));

      case 14:
        context$1$0.next = 35;
        break;

      case 16:
        type = timeoutsObj.type;
        ms = timeoutsObj.ms;

        _logger2['default'].debug('type: ' + type + ', ms: ' + ms);

        context$1$0.t0 = type;
        context$1$0.next = context$1$0.t0 === 'command' ? 22 : context$1$0.t0 === 'implicit' ? 25 : context$1$0.t0 === 'page load' ? 28 : context$1$0.t0 === 'script' ? 31 : 34;
        break;

      case 22:
        context$1$0.next = 24;
        return _regeneratorRuntime.awrap(this.newCommandTimeout(ms));

      case 24:
        return context$1$0.abrupt('break', 35);

      case 25:
        context$1$0.next = 27;
        return _regeneratorRuntime.awrap(this.implicitWaitMJSONWP(ms));

      case 27:
        return context$1$0.abrupt('break', 35);

      case 28:
        context$1$0.next = 30;
        return _regeneratorRuntime.awrap(this.pageLoadTimeoutMJSONWP(ms));

      case 30:
        return context$1$0.abrupt('break', 35);

      case 31:
        context$1$0.next = 33;
        return _regeneratorRuntime.awrap(this.scriptTimeoutMJSONWP(ms));

      case 33:
        return context$1$0.abrupt('break', 35);

      case 34:
        throw new Error('\'' + type + '\' is not supported');

      case 35:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

commands.getTimeouts = function callee$0$0() {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt('return', {
          command: this.newCommandTimeoutMs,
          implicit: this.implicitWaitMs
        });

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// implicit
commands.implicitWaitW3C = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(this.implicitWait(ms));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

commands.implicitWaitMJSONWP = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(this.implicitWait(ms));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

commands.implicitWait = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(this.setImplicitWait(this.parseTimeoutArgument(ms)));

      case 2:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.setImplicitWait = function (ms) {
  this.implicitWaitMs = ms;
  _logger2['default'].debug('Set implicit wait to ' + ms + 'ms');
  if (this.managedDrivers && this.managedDrivers.length) {
    _logger2['default'].debug('Setting implicit wait on managed drivers');
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _getIterator(this.managedDrivers), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var driver = _step.value;

        if (_lodash2['default'].isFunction(driver.setImplicitWait)) {
          driver.setImplicitWait(ms);
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator['return']) {
          _iterator['return']();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }
};

// pageLoad
// eslint-disable-next-line no-unused-vars
commands.pageLoadTimeoutW3C = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        throw new _protocol.errors.NotImplementedError('Not implemented yet for pageLoad.');

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// eslint-disable-next-line no-unused-vars
commands.pageLoadTimeoutMJSONWP = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        throw new _protocol.errors.NotImplementedError('Not implemented yet for pageLoad.');

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// script
// eslint-disable-next-line no-unused-vars
commands.scriptTimeoutW3C = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        throw new _protocol.errors.NotImplementedError('Not implemented yet for script.');

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// eslint-disable-next-line no-unused-vars
commands.scriptTimeoutMJSONWP = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        throw new _protocol.errors.NotImplementedError('Not implemented yet for script.');

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// command
commands.newCommandTimeout = function callee$0$0(ms) {
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        this.setNewCommandTimeout(this.parseTimeoutArgument(ms));

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.setNewCommandTimeout = function (ms) {
  this.newCommandTimeoutMs = ms;
  _logger2['default'].debug('Set new command timeout to ' + ms + 'ms');
  if (this.managedDrivers && this.managedDrivers.length) {
    _logger2['default'].debug('Setting new command timeout on managed drivers');
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = _getIterator(this.managedDrivers), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var driver = _step2.value;

        if (_lodash2['default'].isFunction(driver.setNewCommandTimeout)) {
          driver.setNewCommandTimeout(ms);
        }
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2['return']) {
          _iterator2['return']();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }
  }
};

helpers.clearNewCommandTimeout = function () {
  if (this.noCommandTimer) {
    this.noCommandTimer.cancel();
    this.noCommandTimer = null;
  }
};

helpers.startNewCommandTimeout = function () {
  var _this = this;

  // make sure there are no rogue timeouts
  this.clearNewCommandTimeout();

  // if command timeout is 0, it is disabled
  if (!this.newCommandTimeoutMs) return; // eslint-disable-line curly

  this.noCommandTimer = _appiumSupport.util.cancellableDelay(this.newCommandTimeoutMs);
  this.noCommandTimer.then(function callee$1$0() {
    var errorMessage;
    return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          // eslint-disable-line promise/prefer-await-to-then
          _logger2['default'].warn('Shutting down because we waited ' + (this.newCommandTimeoutMs / 1000 + ' seconds for a command'));
          errorMessage = 'New Command Timeout of ' + (this.newCommandTimeoutMs / 1000 + ' seconds ') + 'expired. Try customizing the timeout using the ' + '\'newCommandTimeout\' desired capability';
          context$2$0.next = 4;
          return _regeneratorRuntime.awrap(this.startUnexpectedShutdown(new Error(errorMessage)));

        case 4:
        case 'end':
          return context$2$0.stop();
      }
    }, null, _this);
  })['catch'](_bluebird2['default'].CancellationError, function () /*err*/{
    // ignore
  });
};

helpers.implicitWaitForCondition = function callee$0$0(condFn) {
  var wrappedCondFn;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    var _arguments = arguments,
        _this2 = this;

    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug('Waiting up to ' + this.implicitWaitMs + ' ms for condition');

        wrappedCondFn = function wrappedCondFn() {
          var args$2$0 = _arguments;
          return _regeneratorRuntime.async(function wrappedCondFn$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                // reset command timeout
                this.clearNewCommandTimeout();

                context$2$0.next = 3;
                return _regeneratorRuntime.awrap(condFn.apply(undefined, args$2$0));

              case 3:
                return context$2$0.abrupt('return', context$2$0.sent);

              case 4:
              case 'end':
                return context$2$0.stop();
            }
          }, null, _this2);
        };

        context$1$0.next = 4;
        return _regeneratorRuntime.awrap((0, _asyncbox.waitForCondition)(wrappedCondFn, {
          waitMs: this.implicitWaitMs, intervalMs: 500, logger: _logger2['default']
        }));

      case 4:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

helpers.parseTimeoutArgument = function (ms) {
  var duration = parseInt(ms, 10);
  if (_lodash2['default'].isNaN(duration) || duration < MIN_TIMEOUT) {
    throw new _protocol.errors.UnknownError('Invalid timeout value \'' + ms + '\'');
  }
  return duration;
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports['default'] = extensions;
