// appium-uiautomator2-driver/build/test/functional/desired.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var uiautomator2ServerLaunchTimeout = process.env.TRAVIS ? 60000 : 20000;
var uiautomator2ServerInstallTimeout = process.env.TRAVIS ? 120000 : 20000;

var GENERIC_CAPS = {
  deviceName: 'Android',
  platformName: 'Android',
  uiautomator2ServerLaunchTimeout: uiautomator2ServerLaunchTimeout,
  uiautomator2ServerInstallTimeout: uiautomator2ServerInstallTimeout,
  automationName: 'uiautomator2'
};

var APIDEMOS_CAPS = _lodash2['default'].defaults({
  app: require.resolve('android-apidemos'),
  appPackage: 'io.appium.android.apis',
  appActivity: '.ApiDemos',
  disableWindowAnimation: true
}, GENERIC_CAPS);

var SCROLL_CAPS = _lodash2['default'].defaults({
  app: require.resolve('android-apidemos'),
  appPackage: 'io.appium.android.apis',
  appActivity: '.view.ScrollView2'
}, GENERIC_CAPS);

var GPS_DEMO_CAPS = _lodash2['default'].defaults({
  app: require.resolve('gps-demo-app'),
  appPackage: 'de.impressive.artworx.tutorials.gps',
  appActivity: '.GPSTest'
}, GENERIC_CAPS);

var BROWSER_CAPS = _lodash2['default'].defaults({
  browserName: 'Chrome'
}, GENERIC_CAPS);

exports.GENERIC_CAPS = GENERIC_CAPS;
exports.APIDEMOS_CAPS = APIDEMOS_CAPS;
exports.GPS_DEMO_CAPS = GPS_DEMO_CAPS;
exports.BROWSER_CAPS = BROWSER_CAPS;
exports.SCROLL_CAPS = SCROLL_CAPS;
