// appium-ios-driver/build/lib/uiauto/uiauto-response.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var MESSAGE_TYPES = ['error', 'no data', 'regular', 'chunk', 'last chunk'];

var UNKNOWN_ERROR = {
  status: 13,
  value: 'Error parsing socket data from instruments'
};

/*
 * Object to contain the data received from the UI Automation system.
 */

var UIAutoResponse = (function () {
  function UIAutoResponse() {
    _classCallCheck(this, UIAutoResponse);

    this.bufferedData = '';
    this.resultBuffer = '';
  }

  _createClass(UIAutoResponse, [{
    key: 'resetBuffer',
    value: function resetBuffer() {
      this.bufferedData = '';
    }
  }, {
    key: 'addData',
    value: function addData(data) {
      this.bufferedData += data;
    }
  }, {
    key: 'finalizeData',
    value: function finalizeData() {
      var data = this.bufferedData;
      this.bufferedData = '';

      // try to figure out what type of data we have, and return it
      var parsedData = undefined;
      try {
        parsedData = {
          type: MESSAGE_TYPES[parseInt(data[0], 10)]
        };
        if (parsedData.type !== 'no data') {
          // format is <one char message type>,<DATA>
          parsedData.result = data.substring(2);
        }
      } catch (err) {
        _logger2['default'].error('Could not parse data from socket: ' + err);
        _logger2['default'].error(data);
        parsedData = {
          type: 'error',
          error: UNKNOWN_ERROR
        };
      }

      return parsedData;
    }
  }, {
    key: 'getResult',
    value: function getResult() {
      var data = this.finalizeData();

      if (!_lodash2['default'].isUndefined(data.result) && data.result !== false) {
        // we have a result, try to make use of it
        if (data.result) {
          _logger2['default'].debug('Got result from instruments: ' + data.result.slice(0, 300));
        } else {
          _logger2['default'].debug('Got null result from instruments');
        }

        if (data.type && data.type.indexOf('chunk') !== -1) {
          // a "chunk" of data, so add to our buffer
          this.resultBuffer += data.result;
          _logger2['default'].debug('Got chunk data, current resultBuffer length: ' + this.resultBuffer.length);
          if (data.type === 'last chunk') {
            _logger2['default'].debug('This is the last data final length: ' + this.resultBuffer.length);
            // this is the last row, unpack and return response
            var result = undefined;
            try {
              result = JSON.parse(this.resultBuffer);
            } catch (err) {
              _logger2['default'].error('Could not parse result buffer: ' + err);
              result = UNKNOWN_ERROR;
            }
            this.resultBuffer = '';
            return result;
          } else {
            _logger2['default'].debug('Not the last chunk, trying to get more');
            return {
              needsMoreData: true
            };
          }
        } else {
          // not a "chunk", so parse and return
          var result = undefined;
          try {
            result = JSON.parse(data.result);
          } catch (err) {
            _logger2['default'].error('Could not parse result buffer: ' + err);
            result = UNKNOWN_ERROR;
          }
          return result;
        }
      } else {
        // we have no result
        return null;
      }
    }
  }]);

  return UIAutoResponse;
})();

exports['default'] = UIAutoResponse;
module.exports = exports['default'];
