// appium-uiautomator2-driver/build/index.js
require('source-map-support').install();

'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _yargs = require('yargs');

var _yargs2 = _interopRequireDefault(_yargs);

var _asyncbox = require('asyncbox');

var _libDriver = require('./lib/driver');

var _libDriver2 = _interopRequireDefault(_libDriver);

var _libServer = require('./lib/server');

var _libServer2 = _interopRequireDefault(_libServer);

var DEFAULT_HOST = "localhost";
exports.DEFAULT_HOST = DEFAULT_HOST;
var DEFAULT_PORT = process.env.TESTOBJECT_E2E_TESTS ? 4723 : 4884;

exports.DEFAULT_PORT = DEFAULT_PORT;
function main() {
  var port, host;
  return _regeneratorRuntime.async(function main$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        port = _yargs2['default'].argv.port || DEFAULT_PORT;
        host = _yargs2['default'].argv.host || _yargs2['default'].argv.address || DEFAULT_HOST;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap((0, _libServer2['default'])(port, host));

      case 4:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

if (require.main === module) {
  (0, _asyncbox.asyncify)(main);
}

exports.AndroidUiautomator2Driver = _libDriver2['default'];
exports.startServer = _libServer2['default'];
exports['default'] = _libDriver2['default'];
