// appium-ios-driver/build/lib/device-log/ios-crash-log.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumSupport = require('appium-support');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var IOSCrashLog = (function () {
  function IOSCrashLog(logDir) {
    _classCallCheck(this, IOSCrashLog);

    this.logDir = logDir || _path2['default'].resolve(process.env.HOME || '/', 'Library', 'Logs', 'DiagnosticReports');
    this.prevLogs = [];
    this.logsSinceLastRequest = [];
  }

  _createClass(IOSCrashLog, [{
    key: 'getCrashes',
    value: function getCrashes() {
      return _regeneratorRuntime.async(function getCrashes$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.prev = 0;
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.glob(this.logDir + '/*.crash'));

          case 3:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](0);

            _logger2['default'].errorAndThrow('There was a problem getting the crash list: ' + context$2$0.t0);

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[0, 6]]);
    }
  }, {
    key: 'filesToJSON',
    value: function filesToJSON(files) {
      return _regeneratorRuntime.async(function filesToJSON$(context$2$0) {
        var _this = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_bluebird2['default'].map(files, function callee$2$0(file) {
              var filename, stat;
              return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
                while (1) switch (context$3$0.prev = context$3$0.next) {
                  case 0:
                    filename = _path2['default'].resolve(this.logDir, file);
                    context$3$0.next = 3;
                    return _regeneratorRuntime.awrap(_appiumSupport.fs.stat(filename));

                  case 3:
                    stat = context$3$0.sent;
                    context$3$0.t0 = stat.ctime.getTime();
                    context$3$0.next = 7;
                    return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(filename, 'utf8'));

                  case 7:
                    context$3$0.t1 = context$3$0.sent;
                    return context$3$0.abrupt('return', {
                      timestamp: context$3$0.t0,
                      level: 'ALL',
                      message: context$3$0.t1
                    });

                  case 9:
                  case 'end':
                    return context$3$0.stop();
                }
              }, null, _this);
            }));

          case 2:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'startCapture',
    value: function startCapture() {
      return _regeneratorRuntime.async(function startCapture$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.getCrashes());

          case 2:
            this.prevLogs = context$2$0.sent;

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'stopCapture',
    value: function stopCapture() {
      return _regeneratorRuntime.async(function stopCapture$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getLogs',

    // needed for consistent API with other logs
    value: function getLogs() {
      var crashFiles, diff;
      return _regeneratorRuntime.async(function getLogs$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.getCrashes());

          case 2:
            crashFiles = context$2$0.sent;
            diff = _lodash2['default'].difference(crashFiles, this.prevLogs, this.logsSinceLastRequest);

            this.logsSinceLastRequest = _lodash2['default'].union(this.logsSinceLastRequest, diff);
            context$2$0.next = 7;
            return _regeneratorRuntime.awrap(this.filesToJSON(diff));

          case 7:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 8:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getAllLogs',
    value: function getAllLogs() {
      var crashFiles, logFiles;
      return _regeneratorRuntime.async(function getAllLogs$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.getCrashes());

          case 2:
            crashFiles = context$2$0.sent;
            logFiles = _lodash2['default'].difference(crashFiles, this.prevLogs);
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(this.filesToJSON(logFiles));

          case 6:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 7:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return IOSCrashLog;
})();

exports.IOSCrashLog = IOSCrashLog;
exports['default'] = IOSCrashLog;
