// appium-ios-driver/build/lib/device-log/ios-performance-log.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var MAX_EVENTS = 5000;

var IOSPerformanceLog = (function () {
  function IOSPerformanceLog(remoteDebugger) {
    var maxEvents = arguments.length <= 1 || arguments[1] === undefined ? MAX_EVENTS : arguments[1];

    _classCallCheck(this, IOSPerformanceLog);

    this.remoteDebugger = remoteDebugger;
    this.maxEvents = parseInt(maxEvents, 10);

    this.timelineEvents = [];
  }

  _createClass(IOSPerformanceLog, [{
    key: 'startCapture',
    value: function startCapture() {
      return _regeneratorRuntime.async(function startCapture$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug('Starting performance (Timeline) log capture');
            this.timelineEvents = [];
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.remoteDebugger.startTimeline(this.onTimelineEvent.bind(this)));

          case 4:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 5:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'stopCapture',
    value: function stopCapture() {
      return _regeneratorRuntime.async(function stopCapture$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug('Stopping performance (Timeline) log capture');
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.remoteDebugger.stopTimeline());

          case 3:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 4:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'onTimelineEvent',
    value: function onTimelineEvent(event) {
      _logger2['default'].debug('Received Timeline event: ' + _lodash2['default'].truncate(JSON.stringify(event)));
      this.timelineEvents.push(event);

      // if we have too many, get rid of the oldest log line
      if (this.timelineEvents.length > this.maxEvents) {
        var removedEvent = this.timelineEvents.shift();
        _logger2['default'].warn('Too many Timeline events, removing earliest: ' + _lodash2['default'].truncate(JSON.stringify(removedEvent)));
      }
    }
  }, {
    key: 'getLogs',
    value: function getLogs() {
      var events;
      return _regeneratorRuntime.async(function getLogs$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            events = this.timelineEvents;

            // flush events
            _logger2['default'].debug('Flushing Timeline events');
            this.timelineEvents = [];

            return context$2$0.abrupt('return', events);

          case 4:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getAllLogs',
    value: function getAllLogs() {
      return _regeneratorRuntime.async(function getAllLogs$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', this.getLogs());

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return IOSPerformanceLog;
})();

exports.IOSPerformanceLog = IOSPerformanceLog;
exports['default'] = IOSPerformanceLog;
