// appium-ios-driver/build/lib/device.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumIosSimulator = require('appium-ios-simulator');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

function checkSimulatorAvailable(opts, sdkVersion, availableDevices) {
  var dString, noDevicesError, sim;
  return _regeneratorRuntime.async(function checkSimulatorAvailable$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(sdkVersion < 7.1)) {
          context$1$0.next = 3;
          break;
        }

        _logger2['default'].debug('Instruments v < 7.1, not checking device string support');
        return context$1$0.abrupt('return');

      case 3:

        _logger2['default'].debug('Checking whether our device string is supported');

        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(getAdjustedDeviceName(opts));

      case 6:
        dString = context$1$0.sent;

        noDevicesError = function noDevicesError() {
          var msg = 'Could not find a device to launch. You requested ' + ('\'' + dString + '\', but the available devices were: ') + JSON.stringify(availableDevices);
          _logger2['default'].errorAndThrow(msg);
        };

        if (!(sdkVersion >= 8)) {
          context$1$0.next = 15;
          break;
        }

        sim = _utils2['default'].getSimForDeviceString(dString, availableDevices);

        if (sim[0] === null || sim[1] === null) {
          noDevicesError();
        }
        _logger2['default'].debug('iOS sim UDID is ' + sim[1]);
        return context$1$0.abrupt('return', sim[1]);

      case 15:
        if (!_lodash2['default'].includes(availableDevices, dString)) {
          noDevicesError();
        }

      case 16:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function getAdjustedDeviceName(opts) {
  return _regeneratorRuntime.async(function getAdjustedDeviceName$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.t0 = opts._adjustedDeviceName;

        if (context$1$0.t0) {
          context$1$0.next = 5;
          break;
        }

        context$1$0.next = 4;
        return _regeneratorRuntime.awrap((0, _appiumIosSimulator.getDeviceString)(opts));

      case 4:
        context$1$0.t0 = context$1$0.sent;

      case 5:
        opts._adjustedDeviceName = context$1$0.t0;
        return context$1$0.abrupt('return', opts._adjustedDeviceName);

      case 7:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

// TODO: what to do with this?
function moveBuiltInApp() {
  return _regeneratorRuntime.async(function moveBuiltInApp$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

/*sim*/
// call sim function once it is in place
function runSimulatorReset(sim, opts, keepApp) {
  return _regeneratorRuntime.async(function runSimulatorReset$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(!opts.reset && !opts.fullReset)) {
          context$1$0.next = 3;
          break;
        }

        _logger2['default'].debug('Reset not set, not ending sim');
        return context$1$0.abrupt('return');

      case 3:

        _logger2['default'].debug('Running ios sim reset flow');

        // The simulator process must be ended before we delete applications.
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(endSimulator(sim));

      case 6:
        if (!opts.fullReset) {
          context$1$0.next = 12;
          break;
        }

        _logger2['default'].debug('Full reset is on, so erasing simulator');
        context$1$0.next = 10;
        return _regeneratorRuntime.awrap(fullResetSimulator(sim));

      case 10:
        context$1$0.next = 15;
        break;

      case 12:
        if (!opts.reset) {
          context$1$0.next = 15;
          break;
        }

        context$1$0.next = 15;
        return _regeneratorRuntime.awrap(resetSimulator(sim, opts, keepApp));

      case 15:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function fullResetSimulator(sim) {
  return _regeneratorRuntime.async(function fullResetSimulator$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug('Cleaning the simulator');

        if (!sim) {
          context$1$0.next = 4;
          break;
        }

        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(sim.clean());

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function resetSimulator(sim, opts, keepApp) {
  return _regeneratorRuntime.async(function resetSimulator$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (sim) {
          context$1$0.next = 2;
          break;
        }

        return context$1$0.abrupt('return');

      case 2:
        // eslint-disable-line curly

        _logger2['default'].debug('Cleaning sim state.');
        context$1$0.prev = 3;
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(clearAppData(sim, opts, keepApp));

      case 6:
        context$1$0.next = 12;
        break;

      case 8:
        context$1$0.prev = 8;
        context$1$0.t0 = context$1$0['catch'](3);

        _logger2['default'].warn(context$1$0.t0);
        _logger2['default'].warn("Could not reset simulator. Leaving as is.");

      case 12:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[3, 8]]);
}

function endSimulator(sim) {
  return _regeneratorRuntime.async(function endSimulator$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (sim) {
          context$1$0.next = 2;
          break;
        }

        return context$1$0.abrupt('return');

      case 2:
        // eslint-disable-line curly

        _logger2['default'].debug('Killing the simulator');
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(sim.shutdown());

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function isolateSimulatorDevice(sim, opts, sdkVersion) {
  return _regeneratorRuntime.async(function isolateSimulatorDevice$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(opts.isolateSimDevice && sdkVersion >= 8)) {
          context$1$0.next = 3;
          break;
        }

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(sim.isolateSim());

      case 3:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function clearAppData(sim, opts, keepApp) {
  return _regeneratorRuntime.async(function clearAppData$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(!keepApp && opts.app && opts.bundleId)) {
          context$1$0.next = 3;
          break;
        }

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(sim.cleanCustomApp(_path2['default'].basename(opts.app), opts.bundleId));

      case 3:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function resetRealDevice(device, opts) {
  var bundleId;
  return _regeneratorRuntime.async(function resetRealDevice$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(opts.bundleId && opts.fullReset)) {
          context$1$0.next = 18;
          break;
        }

        bundleId = opts.bundleId;

        _logger2['default'].debug('Full reset requested. Will try to uninstall the app \'' + bundleId + '\'.');
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(device.isInstalled(bundleId));

      case 5:
        if (context$1$0.sent) {
          context$1$0.next = 8;
          break;
        }

        _logger2['default'].debug('App not installed. No need to uninstall');
        return context$1$0.abrupt('return');

      case 8:
        context$1$0.prev = 8;
        context$1$0.next = 11;
        return _regeneratorRuntime.awrap(device.remove(bundleId));

      case 11:
        context$1$0.next = 17;
        break;

      case 13:
        context$1$0.prev = 13;
        context$1$0.t0 = context$1$0['catch'](8);

        _logger2['default'].error('Could not remove \'' + bundleId + '\' from device');
        throw context$1$0.t0;

      case 17:
        _logger2['default'].debug('Removed ' + bundleId);

      case 18:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[8, 13]]);
}

function runRealDeviceReset(device, opts) {
  return _regeneratorRuntime.async(function runRealDeviceReset$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(opts.reset || opts.fullReset)) {
          context$1$0.next = 7;
          break;
        }

        _logger2['default'].debug("Running ios real device reset flow");

        if (!opts.reset) {
          context$1$0.next = 5;
          break;
        }

        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(resetRealDevice(device, opts));

      case 5:
        context$1$0.next = 8;
        break;

      case 7:
        _logger2['default'].debug("Reset not set, continuing");

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.runSimulatorReset = runSimulatorReset;
exports.isolateSimulatorDevice = isolateSimulatorDevice;
exports.checkSimulatorAvailable = checkSimulatorAvailable;
exports.moveBuiltInApp = moveBuiltInApp;
exports.getAdjustedDeviceName = getAdjustedDeviceName;
exports.endSimulator = endSimulator;
exports.runRealDeviceReset = runRealDeviceReset;
