// appium-ios-driver/build/lib/iwdp.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var _events = require('events');

var _events2 = _interopRequireDefault(_events);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _requestPromise = require('request-promise');

var _requestPromise2 = _interopRequireDefault(_requestPromise);

var _asyncbox = require('asyncbox');

var _appiumSupport = require('appium-support');

var IWDP_CMD = 'ios_webkit_debug_proxy';
var MAX_RETRIES = 10;

var iwdpLogger = _appiumSupport.logger.getLogger('IWDP');

var IWDP = (function (_EventEmitter) {
  _inherits(IWDP, _EventEmitter);

  function IWDP() {
    var webkitDebugProxyPort = arguments.length <= 0 || arguments[0] === undefined ? 27753 : arguments[0];
    var udid = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];

    _classCallCheck(this, IWDP);

    _get(Object.getPrototypeOf(IWDP.prototype), 'constructor', this).call(this);
    this.udid = udid;
    this.attempts = 0;
    this.port = webkitDebugProxyPort;
    this.process = this.createIWDPProcess();
    this.endpoint = 'http://localhost:' + this.port;
  }

  _createClass(IWDP, [{
    key: 'createIWDPProcess',
    value: function createIWDPProcess() {
      var _this = this;

      var process = new _teen_process.SubProcess(IWDP_CMD, ['-c', this.udid + ':' + this.port, '-d']); // (see https://github.com/google/ios-webkit-debug-proxy for reference)
      process.on('exit', function () {
        return _this.onExit();
      });
      process.on('lines-stderr', iwdpLogger.error);
      return process;
    }
  }, {
    key: 'onExit',
    value: function onExit() {
      return _regeneratorRuntime.async(function onExit$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (this.exitRequested) {
              context$2$0.next = 4;
              break;
            }

            this.process = this.createIWDPProcess();
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.start());

          case 4:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'start',
    value: function start() {
      return _regeneratorRuntime.async(function start$(context$2$0) {
        var _this2 = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(++this.attempts > MAX_RETRIES)) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt('return', this.emit('error', new Error('Failed to start IWDP server. Max retry attempts ' + MAX_RETRIES + ' reached')));

          case 2:
            this.exitRequested = false;

            // Throw error if ios_webkit_debug_proxy is not installed
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this.isSupported());

          case 5:
            if (context$2$0.sent) {
              context$2$0.next = 7;
              break;
            }

            _logger2['default'].errorAndThrow('\'ios_webkit_debug_proxy\' not installed on this machine. Try \'brew install ios-webkit-debug-proxy');

          case 7:

            _logger2['default'].debug('Starting ios_webkit_debug_proxy at port ' + this.port + ' on device ' + this.udid);
            context$2$0.next = 10;
            return _regeneratorRuntime.awrap(this.process.start(0));

          case 10:
            context$2$0.prev = 10;
            context$2$0.next = 13;
            return _regeneratorRuntime.awrap((0, _asyncbox.retryInterval)(20, 500, function callee$2$0() {
              return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
                while (1) switch (context$3$0.prev = context$3$0.next) {
                  case 0:
                    context$3$0.next = 2;
                    return _regeneratorRuntime.awrap((0, _requestPromise2['default'])(this.endpoint));

                  case 2:
                    return context$3$0.abrupt('return', context$3$0.sent);

                  case 3:
                  case 'end':
                    return context$3$0.stop();
                }
              }, null, _this2);
            }));

          case 13:
            this.emit('start');
            context$2$0.next = 26;
            break;

          case 16:
            context$2$0.prev = 16;
            context$2$0.t0 = context$2$0['catch'](10);
            context$2$0.prev = 18;
            context$2$0.next = 21;
            return _regeneratorRuntime.awrap(this.process.stop());

          case 21:
            context$2$0.next = 25;
            break;

          case 23:
            context$2$0.prev = 23;
            context$2$0.t1 = context$2$0['catch'](18);

          case 25:
            _logger2['default'].errorAndThrow('Timed out waiting for ios_webkit_debug_proxy to open');

          case 26:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[10, 16], [18, 23]]);
    }
  }, {
    key: 'stop',
    value: function stop() {
      return _regeneratorRuntime.async(function stop$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.attempts = 0;
            this.exitRequested = true;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.process.stop());

          case 4:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 5:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }

    /**
     * Is 'ios_webkit_debug_proxy' available?
     */
  }, {
    key: 'isSupported',
    value: function isSupported() {
      return _regeneratorRuntime.async(function isSupported$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(typeof this.supported !== 'undefined')) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt('return', this.supported);

          case 2:
            context$2$0.prev = 2;
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.which(IWDP_CMD));

          case 5:
            this.supported = true;
            context$2$0.next = 11;
            break;

          case 8:
            context$2$0.prev = 8;
            context$2$0.t0 = context$2$0['catch'](2);

            this.supported = false;

          case 11:
            return context$2$0.abrupt('return', this.supported);

          case 12:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[2, 8]]);
    }
  }]);

  return IWDP;
})(_events2['default']);

exports.IWDP = IWDP;
exports['default'] = IWDP;

// If the process exits and the exit wasn't requested by the API, restart it

// Retry pinging the iwdp server until it's ready
