// appium-ios-driver/build/lib/server.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _appiumBaseDriver = require('appium-base-driver');

var _driver = require('./driver');

var _https = require('https');

var _https2 = _interopRequireDefault(_https);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _requestPromise = require('request-promise');

var _requestPromise2 = _interopRequireDefault(_requestPromise);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _portfinder = require('portfinder');

var _portfinder2 = _interopRequireDefault(_portfinder);

var pem = _bluebird2['default'].promisifyAll(require('pem'));

function startServer(port, host) {
  var driver, router, server;
  return _regeneratorRuntime.async(function startServer$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        driver = new _driver.IosDriver();
        router = (0, _appiumBaseDriver.routeConfiguringFunction)(driver);
        server = (0, _appiumBaseDriver.server)(router, port, host);

        _logger2['default'].info('IosDriver server listening on http://' + host + ':' + port);
        return context$1$0.abrupt('return', server);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function startHttpsServer(port, host) {
  var privateKey, keys, pemCertificate, httpsPort, serverOpts, sslServer;
  return _regeneratorRuntime.async(function startHttpsServer$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(pem.createPrivateKeyAsync());

      case 2:
        privateKey = context$1$0.sent;
        context$1$0.next = 5;
        return _regeneratorRuntime.awrap(pem.createCertificateAsync({ days: 1, selfSigned: true, serviceKey: privateKey.key }));

      case 5:
        keys = context$1$0.sent;
        pemCertificate = keys.certificate;
        context$1$0.next = 9;
        return _regeneratorRuntime.awrap(_portfinder2['default'].getPortPromise());

      case 9:
        httpsPort = context$1$0.sent;
        serverOpts = { key: keys.serviceKey, cert: pemCertificate };
        sslServer = _https2['default'].createServer(serverOpts, function callee$1$0(req, res) {
          return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
            while (1) switch (context$2$0.prev = context$2$0.next) {
              case 0:
                _logger2['default'].debug('Received HTTPS \'' + req.method + '\' request for \'' + req.url + '\'');
                if (req.method === 'OPTIONS') {
                  res.setHeader('Access-Control-Allow-Origin', '*');
                  res.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS');
                  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
                  res.writeHead(200);
                  res.end();
                } else if (req.method === 'POST') {
                  (function () {
                    var body = '';
                    req.on('data', function (data) {
                      body += data;
                    });
                    req.on('end', function callee$3$0() {
                      var httpsUrl, uri, options;
                      return _regeneratorRuntime.async(function callee$3$0$(context$4$0) {
                        while (1) switch (context$4$0.prev = context$4$0.next) {
                          case 0:
                            httpsUrl = _url2['default'].parse(req.url);
                            uri = 'http://' + host + ':' + port + httpsUrl.pathname;

                            _logger2['default'].debug('Passing \'' + body + '\' to \'' + uri + '\'');

                            // now we send to the right place
                            options = {
                              method: 'POST',
                              uri: uri,
                              body: body
                            };
                            context$4$0.next = 6;
                            return _regeneratorRuntime.awrap((0, _requestPromise2['default'])(options));

                          case 6:
                            res.writeHead(200);
                            res.end();

                          case 8:
                          case 'end':
                            return context$4$0.stop();
                        }
                      }, null, this);
                    });
                  })();
                }

              case 2:
              case 'end':
                return context$2$0.stop();
            }
          }, null, this);
        }).listen(httpsPort);
        return context$1$0.abrupt('return', {
          sslServer: sslServer,
          pemCertificate: pemCertificate,
          httpsPort: httpsPort
        });

      case 13:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.startServer = startServer;
exports.startHttpsServer = startHttpsServer;

// Create a random pem certificate

// find a port

// Host an SSL server that uses that certificate
