// appium-ios-driver/build/lib/commands/device.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});
var commands = {},
    helpers = {},
    extensions = {};

// this function is memoized in IOSDriver's constructor
commands.getDevicePixelRatio = function callee$0$0() {
  var screenHeight, screenHeightOfPlusDevices;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(this.getWindowSize());

      case 2:
        screenHeight = context$1$0.sent.height;
        screenHeightOfPlusDevices = 736;
        return context$1$0.abrupt('return', screenHeight === screenHeightOfPlusDevices ? 3.0 : 2.0);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

commands.getViewportRect = function callee$0$0() {
  var scale, statusBarHeight, windowSize;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(this.getDevicePixelRatio());

      case 2:
        scale = context$1$0.sent;
        context$1$0.t0 = Math;
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(this.getStatusBarHeight());

      case 6:
        context$1$0.t1 = context$1$0.sent;
        context$1$0.t2 = scale;
        context$1$0.t3 = context$1$0.t1 * context$1$0.t2;
        statusBarHeight = context$1$0.t0.round.call(context$1$0.t0, context$1$0.t3);
        context$1$0.next = 12;
        return _regeneratorRuntime.awrap(this.getWindowSize());

      case 12:
        windowSize = context$1$0.sent;
        return context$1$0.abrupt('return', {
          left: 0,
          top: statusBarHeight,
          width: windowSize.width * scale,
          height: windowSize.height * scale - statusBarHeight
        });

      case 14:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

commands.getStatusBarHeight = function callee$0$0() {
  var command;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        command = 'UIATarget.localTarget().frontMostApp().statusBar().rect().size.height;';
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(this.uiAutoClient.sendCommand(command));

      case 3:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports['default'] = extensions;

// There is no way to grab scale from UIAutomation. All devices have scale 2.0 except iPhone 6+, iPhone 6s+, iPhone7+ - 3.0.

// status bar height comes in unscaled, so scale it

// ios returns coordinates/dimensions in logical pixels, not device pixels,
// so scale up to device pixels. status bar height is already scaled.
