// appium-ios-driver/build/lib/commands/index.js
'use strict';

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _find = require('./find');

var _find2 = _interopRequireDefault(_find);

var _element = require('./element');

var _element2 = _interopRequireDefault(_element);

var _logging = require('./logging');

var _logging2 = _interopRequireDefault(_logging);

var _gesture = require('./gesture');

var _gesture2 = _interopRequireDefault(_gesture);

var _alert = require('./alert');

var _alert2 = _interopRequireDefault(_alert);

var _execute = require('./execute');

var _execute2 = _interopRequireDefault(_execute);

var _general = require('./general');

var _general2 = _interopRequireDefault(_general);

var _context = require('./context');

var _context2 = _interopRequireDefault(_context);

var _web = require('./web');

var _web2 = _interopRequireDefault(_web);

var _orientation = require('./orientation');

var _orientation2 = _interopRequireDefault(_orientation);

var _fileMovement = require('./file-movement');

var _fileMovement2 = _interopRequireDefault(_fileMovement);

var _navigation = require('./navigation');

var _navigation2 = _interopRequireDefault(_navigation);

var _screenshot = require('./screenshot');

var _screenshot2 = _interopRequireDefault(_screenshot);

var _device = require('./device');

var _device2 = _interopRequireDefault(_device);

var _safari = require('./safari');

var _safari2 = _interopRequireDefault(_safari);

var _coverage = require('./coverage');

var _coverage2 = _interopRequireDefault(_coverage);

var _timeout = require('./timeout');

var _timeout2 = _interopRequireDefault(_timeout);

var commands = {};

var _arr = [_find2['default'], _element2['default'], _logging2['default'], _gesture2['default'], _alert2['default'], _execute2['default'], _general2['default'], _context2['default'], _web2['default'], _orientation2['default'], _fileMovement2['default'], _navigation2['default'], _screenshot2['default'], _device2['default'], _safari2['default'], _coverage2['default'], _timeout2['default']];
for (var _i = 0; _i < _arr.length; _i++) {
  var obj = _arr[_i];
  _Object$assign(commands, obj);
}

var iosCommands = {
  find: _find2['default'],
  element: _element2['default'],
  logging: _logging2['default'],
  gesture: _gesture2['default'],
  alert: _alert2['default'],
  execute: _execute2['default'],
  general: _general2['default'],
  context: _context2['default'],
  web: _web2['default'],
  orientation: _orientation2['default'],
  file: _fileMovement2['default'],
  navigation: _navigation2['default'],
  screenshot: _screenshot2['default'],
  device: _device2['default'],
  safari: _safari2['default'],
  coverage: _coverage2['default'],
  timeout: _timeout2['default']
};

exports['default'] = commands;
exports.commands = commands;
exports.iosCommands = iosCommands;
