// wdio-sauce-service/build/sauce-launch-service.js
'use strict';

Object.defineProperty(exports, '__esModule', {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _sauceConnectLauncher = require('sauce-connect-launcher');

var _sauceConnectLauncher2 = _interopRequireDefault(_sauceConnectLauncher);

var SauceLaunchService = (function () {
    function SauceLaunchService() {
        _classCallCheck(this, SauceLaunchService);
    }

    _createClass(SauceLaunchService, [{
        key: 'onPrepare',

        /**
         * modify config and launch sauce connect
         */
        value: function onPrepare(config) {
            var _this = this;

            if (!config.sauceConnect) {
                return;
            }

            config.host = 'localhost';
            config.port = 4445;

            this.sauceConnectOpts = _extends({
                username: config.user,
                accessKey: config.key
            }, config.sauceConnectOpts);

            return new Promise(function (resolve, reject) {
                return (0, _sauceConnectLauncher2['default'])(_this.sauceConnectOpts, function (err, sauceConnectProcess) {
                    if (err) {
                        return reject(err);
                    }

                    _this.sauceConnectProcess = sauceConnectProcess;
                    resolve();
                });
            });
        }

        /**
         * shut down sauce connect
         */
    }, {
        key: 'onComplete',
        value: function onComplete() {
            var _this2 = this;

            if (!this.sauceConnectProcess) {
                return;
            }

            return new Promise(function (r) {
                return _this2.sauceConnectProcess.close(r);
            });
        }
    }]);

    return SauceLaunchService;
})();

exports['default'] = SauceLaunchService;
module.exports = exports['default'];
