// appium-fake-driver/build/test/alert-tests.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _helpers = require('./helpers');

function alertTests() {
  describe('alerts', function () {
    var driver = undefined;
    before(function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            context$3$0.next = 2;
            return _regeneratorRuntime.awrap((0, _helpers.initSession)(_helpers.DEFAULT_CAPS));

          case 2:
            driver = context$3$0.sent;

          case 3:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    after(function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            context$3$0.next = 2;
            return _regeneratorRuntime.awrap((0, _helpers.deleteSession)());

          case 2:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });

    it('should not work with alerts when one is not present', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            context$3$0.next = 2;
            return _regeneratorRuntime.awrap(driver.alertText().should.eventually.be.rejectedWith(/27/));

          case 2:
            context$3$0.next = 4;
            return _regeneratorRuntime.awrap(driver.alertKeys('foo').should.eventually.be.rejectedWith(/27/));

          case 4:
            context$3$0.next = 6;
            return _regeneratorRuntime.awrap(driver.acceptAlert().should.eventually.be.rejectedWith(/27/));

          case 6:
            context$3$0.next = 8;
            return _regeneratorRuntime.awrap(driver.dismissAlert().should.eventually.be.rejectedWith(/27/));

          case 8:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    it('should get text of an alert', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            context$3$0.next = 2;
            return _regeneratorRuntime.awrap(driver.elementById("AlertButton").click());

          case 2:
            context$3$0.next = 4;
            return _regeneratorRuntime.awrap(driver.alertText());

          case 4:
            context$3$0.sent.should.equal("Fake Alert");

          case 5:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    it('should set the text of an alert', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            context$3$0.next = 2;
            return _regeneratorRuntime.awrap(driver.alertKeys('foo'));

          case 2:
            context$3$0.next = 4;
            return _regeneratorRuntime.awrap(driver.alertText());

          case 4:
            context$3$0.sent.should.equal('foo');

          case 5:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    it('should not do other things while an alert is there', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            context$3$0.next = 2;
            return _regeneratorRuntime.awrap(driver.elementById("nav").click().should.eventually.be.rejectedWith(/26/));

          case 2:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    it.skip('should accept an alert', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            driver.acceptAlert().elementById("nav").click().nodeify();

          case 1:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    it.skip('should not set the text of the wrong kind of alert', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            driver.elementById("AlertButton2").click().alertText().should.eventually.become('Fake Alert 2').alertKeys('foo').should.be.rejectedWith(/12/).nodeify();

          case 1:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
    it.skip('should dismiss an alert', function callee$2$0() {
      return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
        while (1) switch (context$3$0.prev = context$3$0.next) {
          case 0:
            driver.acceptAlert().elementById("nav").click().nodeify();

          case 1:
          case 'end':
            return context$3$0.stop();
        }
      }, null, this);
    });
  });
}

exports['default'] = alertTests;
module.exports = exports['default'];
