// appium-fake-driver/build/test/context-tests.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _helpers = require('./helpers');

function contextTests() {
  describe('contexts, webviews, frames', function callee$1$0() {
    var driver;
    return _regeneratorRuntime.async(function callee$1$0$(context$2$0) {
      while (1) switch (context$2$0.prev = context$2$0.next) {
        case 0:
          driver = undefined;

          before(function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap((0, _helpers.initSession)(_helpers.DEFAULT_CAPS));

                case 2:
                  driver = context$3$0.sent;

                case 3:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          after(function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap((0, _helpers.deleteSession)());

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should get current context', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.currentContext().should.eventually.become('NATIVE_APP'));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should get contexts', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.contexts().should.eventually.become(['NATIVE_APP', 'WEBVIEW_1']));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should not set context that is not there', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.context('WEBVIEW_FOO').should.eventually.be.rejectedWith(/35/));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should set context', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.context('WEBVIEW_1').currentContext().should.eventually.become('WEBVIEW_1'));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should find webview elements in a webview', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.elementByXPath('//*').getTagName().should.eventually.become('html'));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should not switch to a frame that is not there', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.frame('foo').should.eventually.be.rejectedWith(/8/));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should switch to an iframe', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.frame('iframe1').title().should.eventually.become('Test iFrame'));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should switch back to default frame', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.frame(null).title().should.eventually.become('Test Webview'));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should go back to native context', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.context('NATIVE_APP').elementByXPath('//*').getTagName().should.eventually.become('app'));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });
          it('should not set a frame in a native context', function callee$2$0() {
            return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
              while (1) switch (context$3$0.prev = context$3$0.next) {
                case 0:
                  context$3$0.next = 2;
                  return _regeneratorRuntime.awrap(driver.frame('iframe1').should.eventually.be.rejectedWith(/36/));

                case 2:
                case 'end':
                  return context$3$0.stop();
              }
            }, null, this);
          });

        case 13:
        case 'end':
          return context$2$0.stop();
      }
    }, null, this);
  });
}

exports['default'] = contextTests;
module.exports = exports['default'];
