// appium-fake-driver/build/test/helpers.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _wd = require('wd');

var _wd2 = _interopRequireDefault(_wd);

var TEST_HOST = 'localhost';
var TEST_PORT = 4774;
var TEST_APP = _path2['default'].resolve(__dirname, "..", "..", "test", "fixtures", "app.xml");

var DEFAULT_CAPS = { platformName: 'Fake', deviceName: 'Fake', app: TEST_APP };

var driver = undefined;

function initSession(caps) {
  return _regeneratorRuntime.async(function initSession$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        driver = _wd2['default'].promiseChainRemote({ host: TEST_HOST, port: TEST_PORT });
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(driver.init(caps));

      case 3:
        return context$1$0.abrupt('return', driver);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function deleteSession() {
  return _regeneratorRuntime.async(function deleteSession$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.prev = 0;
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(driver.quit());

      case 3:
        context$1$0.next = 7;
        break;

      case 5:
        context$1$0.prev = 5;
        context$1$0.t0 = context$1$0['catch'](0);

      case 7:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[0, 5]]);
}

exports.initSession = initSession;
exports.deleteSession = deleteSession;
exports.TEST_APP = TEST_APP;
exports.TEST_HOST = TEST_HOST;
exports.TEST_PORT = TEST_PORT;
exports.DEFAULT_CAPS = DEFAULT_CAPS;
