// appium-mac-driver/build/lib/driver.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumBaseDriver = require('appium-base-driver');

var _appiumSupport = require('appium-support');

var _appiumForMac = require('./appium-for-mac');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

// Appium instantiates this class

var MacDriver = (function (_BaseDriver) {
  _inherits(MacDriver, _BaseDriver);

  function MacDriver() {
    var opts = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var shouldValidateCaps = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

    _classCallCheck(this, MacDriver);

    _get(Object.getPrototypeOf(MacDriver.prototype), 'constructor', this).call(this, opts, shouldValidateCaps);
    this.jwpProxyActive = false;
    this.opts.address = opts.address || _appiumForMac.DEFAULT_A4M_HOST;
  }

  _createClass(MacDriver, [{
    key: 'createSession',
    value: function createSession() {
      var _len,
          args,
          _key,
          _ref,
          _ref2,
          sessionId,
          caps,
          args$2$0 = arguments;

      return _regeneratorRuntime.async(function createSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (_appiumSupport.system.isMac()) {
              context$2$0.next = 2;
              break;
            }

            throw new Error("AppiumForMac tests only run on the Mac");

          case 2:
            context$2$0.prev = 2;

            for (_len = args$2$0.length, args = Array(_len), _key = 0; _key < _len; _key++) {
              args[_key] = args$2$0[_key];
            }

            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(_get(Object.getPrototypeOf(MacDriver.prototype), 'createSession', this).apply(this, args));

          case 6:
            _ref = context$2$0.sent;
            _ref2 = _slicedToArray(_ref, 2);
            sessionId = _ref2[0];
            caps = _ref2[1];
            context$2$0.next = 12;
            return _regeneratorRuntime.awrap(this.startAppiumForMacSession());

          case 12:
            if (!caps.app) {
              context$2$0.next = 16;
              break;
            }

            _logger2['default'].info('Automatically navigating to app \'' + caps.app + '\'');
            context$2$0.next = 16;
            return _regeneratorRuntime.awrap(this.a4mDriver.sendCommand('/url', 'POST', { url: caps.app }));

          case 16:
            return context$2$0.abrupt('return', [sessionId, caps]);

          case 19:
            context$2$0.prev = 19;
            context$2$0.t0 = context$2$0['catch'](2);
            context$2$0.next = 23;
            return _regeneratorRuntime.awrap(this.deleteSession());

          case 23:
            throw context$2$0.t0;

          case 24:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[2, 19]]);
    }
  }, {
    key: 'startAppiumForMacSession',
    value: function startAppiumForMacSession() {
      return _regeneratorRuntime.async(function startAppiumForMacSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.a4mDriver = new _appiumForMac.AppiumForMac();

            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.a4mDriver.start());

          case 3:
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this.a4mDriver.startSession(this.caps));

          case 5:
            this.proxyReqRes = this.a4mDriver.proxyReqRes.bind(this.a4mDriver);
            // now that everything has started successfully, turn on proxying so all
            // subsequent session requests go straight to/from AppiumForMac
            this.jwpProxyActive = true;

          case 7:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'deleteSession',
    value: function deleteSession() {
      return _regeneratorRuntime.async(function deleteSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug('Deleting AppiumForMac session');

            if (!(this.a4mDriver && this.jwpProxyActive)) {
              context$2$0.next = 7;
              break;
            }

            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.a4mDriver.deleteSession());

          case 4:
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(this.a4mDriver.stop());

          case 6:
            this.a4mDriver = null;

          case 7:
            this.jwpProxyActive = false;
            context$2$0.next = 10;
            return _regeneratorRuntime.awrap(_get(Object.getPrototypeOf(MacDriver.prototype), 'deleteSession', this).call(this));

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'proxyActive',
    value: function proxyActive() {
      // we always have an active proxy to the AppiumForMac server
      return true;
    }
  }, {
    key: 'canProxy',
    value: function canProxy() {
      // we can always proxy to the AppiumForMac server
      return true;
    }
  }, {
    key: 'driverData',
    get: function get() {
      return { A4MPort: this.opts.port };
    }
  }]);

  return MacDriver;
})(_appiumBaseDriver.BaseDriver);

exports.MacDriver = MacDriver;
exports['default'] = MacDriver;
