// appium-base-driver/build/lib/protocol/validators.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumSupport = require('appium-support');

var _basedriverDriver = require("../basedriver/driver");

var _basedriverDriver2 = _interopRequireDefault(_basedriverDriver);

function isNumber(o) {
  return _lodash2['default'].isNumber(o) || !_lodash2['default'].isNaN(parseInt(o, 10)) || !_lodash2['default'].isNaN(parseFloat(o));
}

function msValidator(ms) {
  if (!_lodash2['default'].isNumber(ms) || ms < 0) {
    throw new Error('Wait ms must be a number equal to 0 or greater');
  }
}

var validators = {
  setUrl: function setUrl(url) {
    // either an `xyz://`, `about:`, or `data:` scheme is allowed
    if (!url || !url.match(/^([a-zA-Z0-9_+.-]+:\/\/)|(about:)|(data:)/)) {
      throw new Error('Url or Uri must start with <scheme>://');
    }
  },
  implicitWait: function implicitWait(ms) {
    msValidator(ms);
  },
  asyncScriptTimeout: function asyncScriptTimeout(ms) {
    msValidator(ms);
  },
  timeouts: function timeouts(timeoutsObj) {
    if (timeoutsObj.protocol === _basedriverDriver2['default'].DRIVER_PROTOCOL.W3C) {
      var script = timeoutsObj.script;
      var pageLoad = timeoutsObj.pageLoad;
      var implicit = timeoutsObj.implicit;

      if (_appiumSupport.util.hasValue(script)) {
        msValidator(script);
      }
      if (_appiumSupport.util.hasValue(pageLoad)) {
        msValidator(pageLoad);
      }
      if (_appiumSupport.util.hasValue(implicit)) {
        msValidator(implicit);
      }
    } else {
      var type = timeoutsObj.type;
      var ms = timeoutsObj.ms;

      msValidator(ms);
      if (!_lodash2['default'].includes(['script', 'implicit', 'page load', 'command'], type)) {
        throw new Error('\'' + type + '\' is not a valid timeout type');
      }
    }
  },
  clickCurrent: function clickCurrent(button) {
    if (!(isNumber(button) || _lodash2['default'].isUndefined(button)) || button < 0 || button > 2) {
      throw new Error('Click button must be 0, 1, or 2');
    }
  },
  setNetworkConnection: function setNetworkConnection(type) {
    if (!isNumber(type) || [0, 1, 2, 4, 6].indexOf(type) === -1) {
      throw new Error('Network type must be one of 0, 1, 2, 4, 6');
    }
  }
};

exports.validators = validators;
