// appium-support/build/lib/net.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _fs = require('fs');

var _fs2 = require('./fs');

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _util = require('./util');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _requestPromise = require('request-promise');

var _requestPromise2 = _interopRequireDefault(_requestPromise);

var _jsftp = require('jsftp');

var _jsftp2 = _interopRequireDefault(_jsftp);

function uploadFileToHttp(remoteUrl) {
  var uploadOptions = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
  var response, responseDebugMsg;
  return _regeneratorRuntime.async(function uploadFileToHttp$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug(remoteUrl.protocol + ' upload options: ' + JSON.stringify(uploadOptions));
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap((0, _requestPromise2['default'])(uploadOptions));

      case 3:
        response = context$1$0.sent;
        responseDebugMsg = 'Response code: ' + response.statusCode + '. ' + ('Response body: ' + JSON.stringify(response.body));

        _logger2['default'].debug(responseDebugMsg);

        if (!(response.statusCode >= 400)) {
          context$1$0.next = 8;
          break;
        }

        throw new Error('Cannot upload the recorded media to \'' + remoteUrl.href + '\'. ' + responseDebugMsg);

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function uploadFileToFtp(localFileStream, remoteUrl) {
  var uploadOptions = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];
  return _regeneratorRuntime.async(function uploadFileToFtp$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].debug(remoteUrl.protocol + ' upload options: ' + JSON.stringify(uploadOptions));
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap(new _bluebird2['default'](function (resolve, reject) {
          new _jsftp2['default'](uploadOptions).put(localFileStream, remoteUrl.pathname, function (err) {
            if (err) {
              return reject(err);
            }
            resolve();
          });
        }));

      case 3:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 4:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

/**
 * Uploads the given file to a remote location. HTTP(S) and FTP
 * protocols are supported.
 *
 * @param {string} localPath - The path to a file on the local storage.
 * @param {string} remotePath - The remote URL to upload the file to.
 * @param {Object} uploadOptions - The options set, which depends on the protocol set for remotePath.
 *                                 See https://www.npmjs.com/package/request-promise and
 *                                 https://www.npmjs.com/package/jsftp for more details.
 */
function uploadFile(localPath, remotePath) {
  var uploadOptions = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

  var remoteUrl, _ref, size, timeStarted, timeElapsed;

  return _regeneratorRuntime.async(function uploadFile$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap((0, _fs2.exists)(localPath));

      case 2:
        if (context$1$0.sent) {
          context$1$0.next = 4;
          break;
        }

        throw new Error('\'' + localPath + '\' does not exists or is not accessible');

      case 4:
        remoteUrl = _url2['default'].parse(remotePath);
        context$1$0.next = 7;
        return _regeneratorRuntime.awrap((0, _fs2.stat)(localPath));

      case 7:
        _ref = context$1$0.sent;
        size = _ref.size;

        _logger2['default'].info('Uploading \'' + localPath + '\' of ' + (0, _util.toReadableSizeString)(size) + ' size to \'' + remotePath + '\'...');
        timeStarted = process.hrtime();

        if (!['http:', 'https:'].includes(remoteUrl.protocol)) {
          context$1$0.next = 16;
          break;
        }

        context$1$0.next = 14;
        return _regeneratorRuntime.awrap(uploadFileToHttp(remoteUrl, uploadOptions));

      case 14:
        context$1$0.next = 22;
        break;

      case 16:
        if (!(remoteUrl.protocol === 'ftp:')) {
          context$1$0.next = 21;
          break;
        }

        context$1$0.next = 19;
        return _regeneratorRuntime.awrap(uploadFileToFtp((0, _fs.createReadStream)(localPath), remoteUrl, uploadOptions));

      case 19:
        context$1$0.next = 22;
        break;

      case 21:
        throw new Error('Cannot upload the file at \'' + localPath + '\' to \'' + remotePath + '\'. ' + ('Unsupported remote protocol \'' + remoteUrl.protocol + '\'. ') + 'Only http/https and ftp protocols are supported.');

      case 22:
        timeElapsed = process.hrtime(timeStarted)[0];

        _logger2['default'].info('Uploaded \'' + localPath + '\' of ' + (0, _util.toReadableSizeString)(size) + ' size in ' + timeElapsed + ' second' + (timeElapsed === 1 ? '' : 's'));

      case 24:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.uploadFile = uploadFile;
