// appium-support/build/lib/system.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var SUPPORTED_OSX_VERSIONS = ['10.8', '10.9', '10.10', '10.11', '10.12', '10.13'];

function isWindows() {
  return _os2['default'].type() === 'Windows_NT';
}

function isMac() {
  return _os2['default'].type() === 'Darwin';
}

function isLinux() {
  return !isWindows() && !isMac();
}

function isOSWin64() {
  return process.arch === 'x64' || process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
}

function arch() {
  var _ref, stdout, is64;

  return _regeneratorRuntime.async(function arch$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!(isLinux() || isMac())) {
          context$1$0.next = 8;
          break;
        }

        context$1$0.next = 3;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('uname', ['-m']));

      case 3:
        _ref = context$1$0.sent;
        stdout = _ref.stdout;
        return context$1$0.abrupt('return', stdout.trim() === 'i686' ? '32' : '64');

      case 8:
        if (!isWindows()) {
          context$1$0.next = 11;
          break;
        }

        is64 = this.isOSWin64();
        return context$1$0.abrupt('return', is64 ? '64' : '32');

      case 11:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function macOsxVersion() {
  var stdout, ver;
  return _regeneratorRuntime.async(function macOsxVersion$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        stdout = undefined;
        context$1$0.prev = 1;
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap((0, _teen_process.exec)('sw_vers', ['-productVersion']));

      case 4:
        stdout = context$1$0.sent.stdout.trim();
        context$1$0.next = 10;
        break;

      case 7:
        context$1$0.prev = 7;
        context$1$0.t0 = context$1$0['catch'](1);
        throw new Error('Could not detect Mac OS X Version: ' + context$1$0.t0);

      case 10:
        ver = SUPPORTED_OSX_VERSIONS.find(function (v) {
          return stdout.startsWith(v);
        });

        if (ver) {
          context$1$0.next = 13;
          break;
        }

        throw new Error('Could not detect Mac OS X Version from sw_vers output: \'' + stdout + '\'');

      case 13:
        return context$1$0.abrupt('return', ver);

      case 14:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[1, 7]]);
}

exports.isWindows = isWindows;
exports.isMac = isMac;
exports.isLinux = isLinux;
exports.isOSWin64 = isOSWin64;
exports.arch = arch;
exports.macOsxVersion = macOsxVersion;
