// appium-support/build/lib/tempdir.js
/* This library is originated from temp.js at http://github.com/bruce/node-temp */
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _fs = require('./fs');

var _fs2 = _interopRequireDefault(_fs);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _constants = require('constants');

var _constants2 = _interopRequireDefault(_constants);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var RDWR_EXCL = _constants2['default'].O_CREAT | _constants2['default'].O_TRUNC | _constants2['default'].O_RDWR | _constants2['default'].O_EXCL;

function tempDir() {
  var now, filePath;
  return _regeneratorRuntime.async(function tempDir$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        now = new Date();
        filePath = _path2['default'].join(_os2['default'].tmpdir(), [now.getFullYear(), now.getMonth(), now.getDate(), '-', process.pid, '-', (Math.random() * 0x100000000 + 1).toString(36)].join(''));
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(_fs2['default'].mkdir(filePath));

      case 4:
        return context$1$0.abrupt('return', filePath);

      case 5:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function path(rawAffixes, defaultPrefix) {
  var affixes, name, tempDirectory;
  return _regeneratorRuntime.async(function path$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        affixes = parseAffixes(rawAffixes, defaultPrefix);
        name = [affixes.prefix, affixes.suffix].join('');
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(tempDir());

      case 4:
        tempDirectory = context$1$0.sent;
        return context$1$0.abrupt('return', _path2['default'].join(tempDirectory, name));

      case 6:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

function open(affixes) {
  var filePath, fd;
  return _regeneratorRuntime.async(function open$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return _regeneratorRuntime.awrap(path(affixes, 'f-'));

      case 2:
        filePath = context$1$0.sent;
        context$1$0.prev = 3;
        context$1$0.next = 6;
        return _regeneratorRuntime.awrap(_fs2['default'].open(filePath, RDWR_EXCL, 384));

      case 6:
        fd = context$1$0.sent;
        return context$1$0.abrupt('return', { path: filePath, fd: fd });

      case 10:
        context$1$0.prev = 10;
        context$1$0.t0 = context$1$0['catch'](3);

        _logger2['default'].errorAndThrow(context$1$0.t0);

      case 13:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[3, 10]]);
}

function parseAffixes(rawAffixes, defaultPrefix) {
  var affixes = { prefix: null, suffix: null };
  if (rawAffixes) {
    switch (typeof rawAffixes) {
      case 'string':
        affixes.prefix = rawAffixes;
        break;
      case 'object':
        affixes = rawAffixes;
        break;
      default:
        throw new Error('Unknown affix declaration: ' + affixes);
    }
  } else {
    affixes.prefix = defaultPrefix;
  }
  return affixes;
}

var _static = tempDir();
var openDir = tempDir;

function staticDir() {
  return _regeneratorRuntime.async(function staticDir$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        return context$1$0.abrupt('return', _static);

      case 1:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.open = open;
exports.path = path;
exports.openDir = openDir;
exports.staticDir = staticDir;

// creates a temp directory using the date and a random string

// opens the file in mode 384
